File: vampire_numbers.rb
# http://rosettacode.org/wiki/Vampire_number

# factorization borrowed from http://rosettacode.org/wiki/Factors_of_an_integer#Ruby
def factors(n)
  (1..Math.sqrt(n)).select {|f| n % f == 0}.inject([]) {|f,i| f << n/i; f << i}.sort.uniq
end

def vampire?(number)
  digits = number.to_s.split(//).sort
  return false if digits.count.odd?
  fangs = []
  factors(number).select {|fctr| fctr.to_s.length == digits.count/2 }.each do |factor| 
    f_one, f_two = factor, number/factor
    next if f_one % 10 == 0 && f_two % 10 == 0
    f_one_ary, f_two_ary = f_one.to_s.split(//), f_two.to_s.split(//)
    fangs << [f_one,f_two] if [f_one_ary + f_two_ary].flatten.sort == digits
  end
  return false if fangs.empty?
  fangs.each { |fang| fang.sort! }.uniq
end

found = 0
counter = 1000
until found == 25
  value = vampire?(counter)
  if value != false
    puts "#{counter} : #{value}"
    found += 1
  end
  if counter + 1 == 10_000
    counter = 100_000
  elsif counter + 1 == 1_000_000
    counter = 10_000_00
  else
    counter += 1
  end
end

[16758243290880, 24959017348650, 14593825548650].each do |n|
  val = vampire?(n)
  puts val != false ? "#{n} : #{val}" : "#{val}"
end

File: vampire_numbers.swift
// http://rosettacode.org/wiki/Vampire_number

import Darwin

func quicksort(array:[Int]) -> [Int] {
  if array.count <= 1 {
    return array
  }
  var lesser:[Int] = []
  var greater:[Int] = []
  var equiv:[Int] = []
  let pivot = array[array.count/2]
  for element in array {
    if element < pivot {
      lesser.append(element)
    } else if element > pivot {
      greater.append(element)
    } else {
      equiv.append(element)
    }
  }
  return quicksort(lesser) + equiv + quicksort(greater)
}

func removeDuplicates(array:[Int]) -> [Int] {
  var mutableArray:[Int] = []
  for var i = 0; i < array.count; i++ {
    let element = array[i]
    if !(contains(mutableArray,element)) {
      mutableArray.append(element)
    }
  }
  return mutableArray
}

func split(number:Int,array:[Int]) -> [Int] {
  var mutableArray = array
  if number < 10 {
    mutableArray.append(number)
    return mutableArray
  } else {
    mutableArray.append(number % 10)
    return split(number/10,mutableArray)
  }
}

func factors(n:Float) -> [Int] {
  let number = Int(n)
  let root = Int(sqrt(n))
  var values:[Int] = []
  var factors:[Int] = []
  for var i = 1; i <= root; i++ {
    if number % i == 0 {
      values.append(i)
    }
  }
  for value in values {
    factors.append(number/value)
    factors.append(value)
  }
  return removeDuplicates(quicksort(factors))
}

func vampire(n:Int) -> [[Int]] {
  let array:[Int] = []
  let digits = quicksort(split(n,array))
  var fangs:[[Int]] = []
  let nfactors = factors(Float(n))
  for nfactor in nfactors {
    let fone = nfactor
    let ftwo = n/nfactor
    if !(fone % 10 == 0 && ftwo % 10 == 0) && (split(fone,array)).count == (split(ftwo,array)).count {
      let val = quicksort(split(fone,array) + split(ftwo,array)) 
      if val == digits {
        let pair = [fone,ftwo]
        fangs.append(pair)
      }
    } 
  }
  return fangs
}

var found = 0
var counter = 1000
while found < 25 {
  let val = vampire(counter)
  if val.count > 0 {
    println("\(counter) : \(val)")
    found++
  }
  if counter + 1 == 10_000 {
    counter = 100_000
  } else if counter + 1 == 1_000_000 {
    counter = 10_000_000
  } else {
    counter++
  }
}